slides/aml-19-040418-working-with-text-data/index.html

```diff
@@ -0,0 +1,1020 @@
+```
+Vocabulary size 1-4gram: 7815528
+```
+]
+
+.smaller[
+- More than 20x more 4-grams!
+]
+
+---
+#Stop-word impact on bi-grams
+
+.tiny-code[
+```python
+cv = CountVectorizer(ngram_range=(1, 2), min_df=4).fit(text_train_sub)
+print("Vocabulary size (1, 2), min_df=4: {}".format(len(cv.vocabulary_)))
+cv = CountVectorizer(ngram_range=(1, 2), min_df=4, stop_words="english").fit(text_train_sub)
+print("Vocabulary size (1, 2), stopwords, min_df=4: {}".format(len(cv.vocabulary_)))
+```
+```
+Vocabulary size (1, 2), min_df=4: 155354
+Vocabulary size (1, 2), stopwords, min_df=4: 81085
+```
+]
+
+---
+# Stop-word impact on 4-grams
+
+
+
+.tiny-code[
+```python
+cv4 = CountVectorizer(ngram_range=(4, 4), min_df=4).fit(text_train_sub)
+cv4sw = CountVectorizer(ngram_range=(4, 4), min_df=4, stop_words="english").fit(text_train_sub)
+print(len(cv4.get_feature_names()))
+print(len(cv4sw.get_feature_names()))```
+```
+31585
+369
+```
+]
+---
+- Stopwords included
+.center[
+![:scale 75%](images/stopwords_1.png)
+]
+
+- stopwords removed (slightly worse)
+.center[
+![:scale 75%](images/stopwords_2.png)
+]
+
+---
+
+.tiny-code[
+```python
+my_stopwords = set(ENGLISH_STOP_WORDS)
+my_stopwords.remove("well")
+my_stopwords.remove("not")
+my_stopwords.add("ve")
+```
+```python
+vect3msw = CountVectorizer(ngram_range=(1, 3), min_df=4, stop_words=my_stopwords)
+X_train3msw = vect3msw.fit_transform(text_train_sub)
+lr3msw = LogisticRegressionCV().fit(X_train3msw, y_train_sub)
+X_val3msw = vect3msw.transform(text_val)
+lr3msw.score(X_val3msw, y_val)
+```
+```
+0.88319999999999999
+```
+```python
+plt.figure(figsize=(15, 4))
+plot_important_features(lr3msw.coef_.ravel(), np.array(vect3msw.get_feature_names()), top_n=40)
+```
+
+]
+
+.center[
+![:scale 90%](images/stopwords_2.png)
+]
+---
+
+# Tf-idf rescaling
+
+$$ tf\text{-}idf(t,d) = tf(t,d)\cdot idf(t)$$
+
+$$ idf(t) = log\frac{1+n_d}{1+df(d,t)} + 1$$
+
+Where, $n_d$ is the total number of documents and $df(d,t)$ is the number of documents containing term $t$
+
+* Emplasizes "rare" words - "soft stop word removal"
+* Slightly non-standard smoothing (many +1s)
+* By default also L2 normalisation!
+
+---
+# TfidfVectorizer, TfidfTransformer
+
+.smaller[
+```python
+from sklearn.feature_extraction.text import TfidfVectorizer, TfidfTransformer
+```
+```python
+malory_tfidf = TfidfVectorizer().fit_transform(malory)
+malory_tfidf.toarray()
+```
+```
+array([[ 0.41 ,  0.   ,  0.576,  0.   ,  0.   ,  0.   ,  0.576,  0.41 ],
+       [ 0.318,  0.447,  0.   ,  0.447,  0.447,  0.447,  0.   ,  0.318]])
+```
+```python
+malory_tfidf = make_pipeline(CountVectorizer(), TfidfTransformer()).fit_transform(malory)
+malory_tfidf.toarray()
+```
+```
+array([[ 0.41 ,  0.   ,  0.576,  0.   ,  0.   ,  0.   ,  0.576,  0.41 ],
+       [ 0.318,  0.447,  0.   ,  0.447,  0.447,  0.447,  0.   ,  0.318]])
+```
+
+]
+
+---
+class: middle
+
+# Character n-grams
+
+---
+#Principle
+<br>
+<br>
+
+.center[
+![:scale 100%](images/do_you_want_ants.png)
```

Comment:
hm this was an animation with a window of size 3 sliding over the text...  I realize it's not in the PDF... do you think you could get it from the odp? we need to have 4 different images here, I guess...